File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseApi.java
package br.com.jonathanzanella.myexpenses.expense;

import android.support.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;

import java.io.IOException;
import java.util.List;

import br.com.jonathanzanella.myexpenses.MyApplication;
import br.com.jonathanzanella.myexpenses.log.Log;
import br.com.jonathanzanella.myexpenses.server.Server;
import br.com.jonathanzanella.myexpenses.sync.UnsyncModel;
import br.com.jonathanzanella.myexpenses.sync.UnsyncModelApi;
import retrofit2.Call;
import retrofit2.Response;

public class ExpenseApi implements UnsyncModelApi<Expense> {
    private static final String LOG_TAG = ExpenseApi.class.getSimpleName();
    private final ExpenseRepository expenseRepository;
    private ExpenseInterface expenseInterface;

    public ExpenseApi(ExpenseRepository expenseRepository) {
        this.expenseRepository = expenseRepository;
    }

    private ExpenseInterface getInterface() {
        if(expenseInterface == null)
            expenseInterface = new Server(MyApplication.getContext()).expenseInterface();
        return expenseInterface;
    }

    @Override
    @Nullable
    public List<Expense> index() {
        Call<List<Expense>> caller = getInterface().index(greaterUpdatedAt());

        try {
            Response<List<Expense>> response = caller.execute();
            if(response.isSuccessful()) {
                return response.body();
            } else {
                Log.error(LOG_TAG, "Index request error: " + response.message());
            }
        } catch (IOException e) {
            Log.error(LOG_TAG, "Index request error: " + e.getMessage());
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public void save(UnsyncModel model) {
        Expense expense = (Expense) model;
        Call<Expense> caller;
        if(StringUtils.isNotEmpty(expense.getServerId()))
            caller = getInterface().update(expense.getServerId(), expense);
        else
            caller = getInterface().create(expense);

        try {
            Response<Expense> response = caller.execute();
            if(response.isSuccessful()) {
                expenseRepository.syncAndSave(response.body());
                Log.info(LOG_TAG, "Updated: " + expense.getData());
            } else {
                Log.error(LOG_TAG, "Save request error: " + response.message() + " uuid: " + expense.getUuid());
            }
        } catch (IOException e) {
            Log.error(LOG_TAG, "Save request error: " + e.getMessage() + " uuid: " + expense.getUuid());
            e.printStackTrace();
        }
    }

    @Override
    public void syncAndSave(UnsyncModel unsync) {
        if(!(unsync instanceof Expense))
            throw new UnsupportedOperationException("UnsyncModel is not a Expense");
        expenseRepository.syncAndSave((Expense) unsync);
    }

    @Override
    public List<Expense> unsyncModels() {
        return expenseRepository.unsync();
    }

    @Override
    public long greaterUpdatedAt() {
        return expenseRepository.greaterUpdatedAt();
    }
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/expense/ExpenseApi.kt
package br.com.jonathanzanella.myexpenses.expense

import br.com.jonathanzanella.myexpenses.MyApplication
import br.com.jonathanzanella.myexpenses.server.Server
import br.com.jonathanzanella.myexpenses.sync.UnsyncModel
import br.com.jonathanzanella.myexpenses.sync.UnsyncModelApi
import org.apache.commons.lang3.StringUtils
import timber.log.Timber
import java.io.IOException

class ExpenseApi(private val expenseRepository: ExpenseRepository) : UnsyncModelApi<Expense> {
    private val expenseInterface : ExpenseInterface by lazy {
        Server(MyApplication.getContext()).expenseInterface()
    }

    override fun index(): List<Expense> {
        val lastUpdatedAt = greaterUpdatedAt()
        Timber.tag("AccountApi.index with lastUpdatedAt: $lastUpdatedAt")
        val caller = expenseInterface.index(lastUpdatedAt)

        return try {
            val response = caller.execute()
            if (response.isSuccessful) {
                response.body()
            } else {
                Timber.e("Index request error: " + response.message())
                ArrayList()
            }
        } catch (e: IOException) {
            Timber.e("Index request error: " + e.message)
            e.printStackTrace()
            ArrayList()
        }
    }

    override fun save(model: UnsyncModel) {
        val expense = model as Expense
        val caller = if (StringUtils.isNotEmpty(expense.serverId))
            expenseInterface.update(expense.serverId!!, expense)
        else
            expenseInterface.create(expense)

        try {
            val response = caller.execute()
            if (response.isSuccessful) {
                expenseRepository.syncAndSave(response.body())
                Timber.i("Updated: " + expense.getData())
            } else {
                Timber.e("Save request error: " + response.message() + " uuid: " + expense.uuid)
            }
        } catch (e: IOException) {
            Timber.e("Save request error: " + e.message + " uuid: " + expense.uuid)
            e.printStackTrace()
        }
    }

    override fun syncAndSave(unsync: UnsyncModel) {
        if (unsync !is Expense)
            throw UnsupportedOperationException("UnsyncModel is not a Expense")
        expenseRepository.syncAndSave(unsync)
    }

    override fun unsyncModels(): List<Expense> {
        return expenseRepository.unsync()
    }

    override fun greaterUpdatedAt(): Long {
        return expenseRepository.greaterUpdatedAt()
    }
}
